libtiledbsoma/include/tiledbsoma/arrow_adapter.h

```diff
@@ -2,7 +2,9 @@
 #define ARROW_ADAPTER_H
 
 #include <tiledbsoma/tiledbsoma>
+#ifndef ARROW_SCHEMA_AND_ARRAY_DEFINED
```

Comment:
If this is a problem, I suspect we have an issue with public vs private symbols. @nguyenv / @gspowley let's make sure we are not exporting symbols outside the public API.

Reply:
It's not a symbol issue. I just balks at 'data structure already defined'.  Ie if I remove the define here we end a sudden death as in:

```r
edd@rob:~/git/tiledb-soma/apis/r(de/sc-25958/refactor_arrow)$ install.r
* installing *source* package found in current working directory ...
* installing *source* package ‘tiledbsoma’ ...
** using staged installation
** static library present
** updated src/Makevars for system library via pkg-config
** libs
ccache g++  -std=gnu++17 -I"/usr/share/R/include" -DNDEBUG -I. -I../inst/include/ -I/usr/local/include  -DR_BUILD  -Iinclude -I'/usr/local/lib/R/site-library/Rcpp/include' -I'/usr/local/lib/R/site-library/RcppSpdlog/include'    -fpic  -g -O3 -Wall -pipe  -Wno-parentheses -Wno-ignored-attributes -Wno-unused-local-typedefs -Wno-deprecated-declarations -Wno-unused-function  -c rinterface.cpp -o rinterface.o
In file included from include/tiledbsoma/arrow_adapter.h:6,
                 from include/tiledbsoma/tiledbsoma:37,
                 from rinterface.cpp:6:
include/tiledbsoma/carrow.h:16:8: error: redefinition of ‘struct ArrowSchema’
   16 | struct ArrowSchema {
      |        ^~~~~~~~~~~
In file included from rinterface.cpp:2:
./nanoarrow.h:83:8: note: previous definition of ‘struct ArrowSchema’
   83 | struct ArrowSchema {
      |        ^~~~~~~~~~~
include/tiledbsoma/carrow.h:32:8: error: redefinition of ‘struct ArrowArray’
   32 | struct ArrowArray {
      |        ^~~~~~~~~~
./nanoarrow.h:99:8: note: previous definition of ‘struct ArrowArray’
   99 | struct ArrowArray {
      |        ^~~~~~~~~~
make: *** [/usr/lib/R/etc/Makeconf:178: rinterface.o] Error 1
ERROR: compilation failed for package ‘tiledbsoma’
* removing ‘/usr/local/lib/R/site-library/tiledbsoma’
* restoring previous ‘/usr/local/lib/R/site-library/tiledbsoma’
Warning message:
In install.packages(pkgs = f, lib = lib, repos = if (isMatchingFile(f)) NULL else repos) :
  installation of package ‘.’ had non-zero exit status
edd@rob:~/git/tiledb-soma/apis/r(de/sc-25958/refactor_arrow)$ 
```

Recall that all this is about using the _C interface to Arrow_ data so we are not accessing any functions or linkable symbols (unless my coffee is too weak and I miss something here).

Reply:
There are alternatives.  We can protect `struct ArrowSchema` and `struc ArrowArray` on this side inside a namespace (and using via `cmake` may even automate this).  We could possibly include `nanoarrow` in `libtiledbsoma` (it has a lot of nice and simple (fast to compile) C functionality).

Reply:
@Shelnutt2 @nguyenv @gspowley Please do let know if you see issues here.  The PR is otherwise approved and ready to be merged.